apps/prairielearn/src/ee/pages/studentCourseInstanceUpgrade/studentCourseInstanceUpgrade.ts

```diff
@@ -17,7 +44,222 @@ router.get(
       return;
     }
 
-    res.send(StudentCourseInstanceUpgrade({ resLocals: res.locals }));
+    const institution = InstitutionSchema.parse(res.locals.institution);
+    const course = CourseSchema.parse(res.locals.course);
+    const course_instance = CourseInstanceSchema.parse(res.locals.course_instance);
+    const user = UserSchema.parse(res.locals.authn_user);
+
+    const planGrants = await getPlanGrantsForPartialContexts({
+      institution_id: institution.id,
+      course_instance_id: course_instance.id,
+      user_id: user.user_id,
+    });
+    const requiredPlans = await getRequiredPlansForCourseInstance(res.locals.course_instance.id);
+    const missingPlans = getMissingPlanGrants(planGrants, requiredPlans);
+
+    // TODO: fetch pricing information from Stripe API; cache it too.
+
+    res.send(
+      StudentCourseInstanceUpgrade({
+        course,
+        course_instance,
+        missingPlans,
+        resLocals: res.locals,
+      }),
+    );
+  }),
+);
+
+const UpgradeBodySchema = z.object({
+  terms_agreement: z.literal('1').optional(),
+  unsafe_plan_names: z.string(),
+});
+
+// Only a subset of all plans are allowed to be paid for on this page.
+const PlanNamesSchema = z.array(z.enum(['basic', 'compute']));
+
+router.post(
+  '/',
+  asyncHandler(async (req, res) => {
+    if (req.body.__action === 'upgrade') {
+      console.log(res.locals);
+      const institution = InstitutionSchema.parse(res.locals.institution);
+      const course = CourseSchema.parse(res.locals.course);
+      const course_instance = CourseInstanceSchema.parse(res.locals.course_instance);
+      const user = UserSchema.parse(res.locals.authn_user);
+
+      const body = UpgradeBodySchema.parse(req.body);
+
+      if (!body.terms_agreement) {
+        throw error.make(400, 'You must agree to the terms and conditions.');
+      }
+
+      const rawPlanNames = body.unsafe_plan_names.split(',');
+      const planNames = PlanNamesSchema.parse(rawPlanNames);
+
+      // TODO: should we use lookup keys instead? See
+      // https://stripe.com/docs/products-prices/manage-prices#lookup-keys
+      // Unfortunately, these can't be set or modified from the Stripe console,
+      // so they aren't as useful as they could be.
+      const lineItems: Stripe.Checkout.SessionCreateParams.LineItem[] = [];
+      if (planNames.includes('basic')) {
+        lineItems.push({
+          price: config.stripePriceIds.basic,
+          quantity: 1,
+        });
+      }
+      if (planNames.includes('compute')) {
+        lineItems.push({
+          price: config.stripePriceIds.compute,
+          quantity: 1,
+        });
+      }
```

Comment:
This assumes that `basic` and `compute` pricing is completely independent - that is, that `basic` costs $10 a-la-carte, `compute` costs $6 a-la-carte, and together they cost $16. This may be a bad assumption - we might want to charge $10 for a-la-carte compute?

Reply:
That said, I'm more than happy to do the naive thing for now and treat them as totally independent.